edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesRuntimeInfoProvider.cs

```diff
@@ -0,0 +1,324 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.IO;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using k8s;
+    using k8s.Models;
+    using Microsoft.Azure.Devices.Edge.Agent.Core;
+    using AgentDocker = Microsoft.Azure.Devices.Edge.Agent.Docker;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+    using Microsoft.Extensions.Logging;
+    using Microsoft.Rest;
+
+
+    public class KubernetesRuntimeInfoProvider : IKubernetesOperator, IRuntimeInfoProvider, INotifyPropertyChanged
+    {
+        readonly string deviceNamespace;
+        readonly IKubernetes client;
+        Option<Watcher<V1Pod>> podWatch;
+        readonly Dictionary<string, ModuleRuntimeInfo> moduleRuntimeInfos;
+        readonly AsyncLock moduleLock;
+
+        public KubernetesRuntimeInfoProvider(string deviceNamespace, IKubernetes client)
+        {
+            this.deviceNamespace = Preconditions.CheckNonWhiteSpace(deviceNamespace,nameof(deviceNamespace));
+            this.client = Preconditions.CheckNotNull(client, nameof(client));
+
+            this.podWatch = Option.None<Watcher<V1Pod>>();
+            this.moduleRuntimeInfos = new Dictionary<string, ModuleRuntimeInfo>();
+            this.moduleLock = new AsyncLock();
+        }
+
+        private async Task ListPodComplete(Task<HttpOperationResponse<V1PodList>> podListRespTask)
+        {
+            if (podListRespTask != null)
+            {
+                HttpOperationResponse<V1PodList> podListResp = await podListRespTask;
+                if (podListResp != null)
+                {
+                    this.podWatch = Option.Some(
+                        podListResp.Watch<V1Pod>(
+                            onEvent: async (type, item) =>
+                            {
+                                try
+                                {
+                                    await this.WatchPodEventsAsync(type, item);
+                                }
+                                catch (Exception ex) when (!ex.IsFatal())
+                                {
+                                    Events.ExceptionInPodWatch(ex);
+                                }
+                            },
+                            onClosed: () =>
+                            {
+                                Events.PodWatchClosed();
+
+                                // get rid of the current pod watch object since we got closed
+                                this.podWatch.ForEach(watch => watch.Dispose());
+                                this.podWatch = Option.None<Watcher<V1Pod>>();
+
+                                // kick off a new watch
+                                this.client.ListNamespacedPodWithHttpMessagesAsync(deviceNamespace, watch: true).ContinueWith(this.ListPodComplete);
```

Comment:
I really don't like the use of continuewith. I think it makes the code less readable and harder to debug. Why can't this be:
 
await this.client.ListNamespacedPodWithHttpMessagesAsync(deviceNamespace, watch: true);
this.ListPodComplete

Reply:
`await this.client.ListNamespacedPodWithHttpMessagesAsync()` blocks until a pod exists in the namespace, and we don't want to block the calling function.  If there's a classier way to do this, I would love to know it.